src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.Helpers.cs

```diff
@@ -36,7 +36,7 @@ private void ValidateWritingProperty()
         {
             if (!_options.SkipValidation)
             {
-                if (!_inObject || _tokenType == JsonTokenType.PropertyName)
+                if (!_inObject || _tokenType == JsonTokenType.PropertyName || _tokenType == JsonTokenType.StringSegment)
```

Comment:
Writing partial string values should not be possible for properties, only string values.

Reply:
This was done to stop someone calling `WritePropertyName` when a segment is being built. The alternative is to implicitly end the string value for these cases instead, which essentially makes `isFinalSegment` a redundant concept. Is it worth revisiting the proposed contract to remove `isFinalSegment`?

Reply:
If not, I should probably introduce a new `ExceptionResource` along the lines of "CannotWriteWithinStringSegment" and make sure it's used in all relevant places for clarity.